zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1098 @@
+        (a, b) => match a.cmp(b) {
+            Ordering::Less => Dominance::from(insert),
+            Ordering::Equal => Dominance::Equal,
+            Ordering::Greater => Dominance::from(!insert),
+        },
+    }
+}
+
+/// In the comments for each alternative, `()` represents the left range and `[]` represents the right range.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum RangeOrdering {
+    /// `(   ) [   ]`
+    LeftFirstDisjoint,
+    /// `( [   ) ]`
+    LeftFirstOverlap,
+    /// `[ (   ) ]`
+    LeftContained,
+    /// ```text
+    /// (   )
+    /// [   ]
+    /// ```
+    Equal,
+    /// `( [   ] )`
+    RightContained,
+    /// `[ (   ] )`
+    RightFirstOverlap,
+    /// `[   ] (   )`
+    RightFirstDisjoint,
+}
+
+impl RangeOrdering {
+    fn cmp<A: Ord>(a: &Range<A>, b: &Range<A>) -> Self {
+        use Ordering::*;
+        assert!(a.start <= a.end && b.start <= b.end);
+        match (a.start.cmp(&b.start), a.end.cmp(&b.end)) {
+            _ if a.end <= b.start => RangeOrdering::LeftFirstDisjoint,
+            _ if b.end <= a.start => RangeOrdering::RightFirstDisjoint,
+            (Less, Less) => RangeOrdering::LeftFirstOverlap,
+            (Equal, Less) | (Greater, Less) | (Greater, Equal) => RangeOrdering::LeftContained,
+            (Equal, Equal) => RangeOrdering::Equal,
+            (Equal, Greater) | (Less, Greater) | (Less, Equal) => RangeOrdering::RightContained,
+            (Greater, Greater) => RangeOrdering::RightFirstOverlap,
+        }
+    }
+}
+
+enum Joined {
+    One(ScanRange),
+    Two(ScanRange, ScanRange),
+    Three(ScanRange, ScanRange, ScanRange),
+}
+
+fn join_nonoverlapping(left: ScanRange, right: ScanRange) -> Joined {
+    assert!(left.block_range().end <= right.block_range().start);
+
+    if left.block_range().end == right.block_range().start {
+        if left.priority() == right.priority() {
+            Joined::One(ScanRange::from_parts(
+                left.block_range().start..right.block_range().end,
+                left.priority(),
+            ))
+        } else {
+            Joined::Two(left, right)
+        }
+    } else {
+        // there is a gap that will need to be filled
+        let gap = ScanRange::from_parts(
+            left.block_range().end..right.block_range().start,
+            ScanPriority::Historic,
+        );
+
+        match join_nonoverlapping(left, gap) {
+            Joined::One(left) => join_nonoverlapping(left, right),
+            Joined::Two(left, gap) => match join_nonoverlapping(gap, right) {
+                Joined::One(right) => Joined::Two(left, right),
+                Joined::Two(gap, right) => Joined::Three(left, gap, right),
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+fn insert(current: ScanRange, to_insert: ScanRange) -> Joined {
+    fn join_overlapping(left: ScanRange, right: ScanRange, insert: Insert) -> Joined {
```

Comment:
As a matter of style, this doesn't capture anything from the outer scope so I would put it at top-level. (I know it's only used inside `insert` but that doesn't really matter.) Non-blocking.